### PR TITLE
Ea 2602 minimal validation

### PR DESCRIPTION
**Pull request by SrdjanStevanetic**

### CODE REVIEW

File: entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsCleaner.java [hunk 1]
```diff
@@ -33,7 +33,7 @@ public EntityFieldsCleaner(LanguageCodes emLanguageCodes) {
 	this.emLanguageCodes = emLanguageCodes;
     }
 
-    public void initialize(ValidEntityFields constraint) {
+    public void initialize(EntityFieldsCompleteValidatorInterface constraint) {
```
Comment (gsergiu) [inline entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsCleaner.java]: I think that this method is not used in the Cleaner class, it should be used only in Validator class. Please remove

File: entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsCompleteValidator.java [hunk 2]
```diff
@@ -194,17 +200,15 @@ boolean validateDateListField(ConstraintValidatorContext context, Field field, L
...
-            return false;
-        }
-
-    	try {
+        try {
             //LocalDate.parse(fieldValue.toString(), DateTimeFormatter.ofPattern(java.time.format.DateTimeFormatter.ISO_DATE.toString()).withResolverStyle(ResolverStyle.STRICT));
-            LocalDate.parse(fieldValue, java.time.format.DateTimeFormatter.ISO_DATE);
+            if(fieldValue.contains("T")) {
+                //ISO Date time format
+                LocalDate.parse(fieldValue, java.time.format.DateTimeFormatter.ISO_DATE_TIME);
```
Comment (gsergiu) [inline entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsCompleteValidator.java:207]: this seems to be an old version of the class. The develop code was updated, please merge the develop branch

File: entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsMinimalValidator.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+		
+		boolean returnValue=true;
+		
+		List<Field> entityFields = EntityUtils.getAllFields(entity.getClass());
+		
+		try {
+			for (Field field : entityFields) {
+				Object fieldValue = entity.getFieldValue(field);
+				if (fieldValue==null) { 
+					addConstraint(context, "The mandatory field: "+field.getName()+" cannot be NULL.");
```
Comment (gsergiu) [inline entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsMinimalValidator.java:38]: this contrain must be applied only for fields that are mandatory, based on the field cardinality 1..1 or 1..*

File: entity-management-web/src/main/java/eu/europeana/entitymanagement/batch/processor/EntityDereferenceProcessor.java [hunk 4]
```diff
@@ -25,11 +33,13 @@
     private static final Logger logger = LogManager.getLogger(EntityDereferenceProcessor.class);
     private final MetisDereferenceService dereferenceService;
     private final EntityComparator entityComparator;
+    private final ValidatorFactory emValidatorFactory;
 
     @Autowired
-    public EntityDereferenceProcessor(MetisDereferenceService dereferenceService) {
+    public EntityDereferenceProcessor(MetisDereferenceService dereferenceService, ValidatorFactory emValidatorFactory) {
```
Comment (gsergiu) [inline entity-management-web/src/main/java/eu/europeana/entitymanagement/batch/processor/EntityDereferenceProcessor.java:39]: the minimal validation must not be integrated into the dereferencing step, but as own functionality in the EntityUpdateProcessor see step 4 in the specs https://docs.google.com/document/d/16k9PcCMFwl2LXjnnzotZRPc-QqM-Ar1D0VELHt4t_hA/edit#

File: entity-management-web/src/main/java/eu/europeana/entitymanagement/batch/processor/EntityDereferenceProcessor.java [hunk 5]
```diff
@@ -101,4 +117,11 @@ private boolean isEuropeanaProxyUpdated(EntityRecord entityRecord){
   private boolean metisResponseMatchesExternalProxy(EntityRecord entityRecord, Entity metisResponse){
       return entityComparator.compare(entityRecord.getExternalProxy().getEntity(), metisResponse) == 0;
     }
+  
+  private void validateConstraints(Entity entity) throws EntityValidationException  {
```
Comment (gsergiu) [inline entity-management-web/src/main/java/eu/europeana/entitymanagement/batch/processor/EntityDereferenceProcessor.java]: the minimal validation must not be integrated into the dereferencing step, but as own functionality in the EntityUpdateProcessor see step 4 in the specs https://docs.google.com/document/d/16k9PcCMFwl2LXjnnzotZRPc-QqM-Ar1D0VELHt4t_hA/edit#

File: entity-management-web/src/test/java/eu/europeana/entitymanagement/validation/EntityFieldsValidatorTest.java [hunk 6]
```diff
@@ -43,33 +45,54 @@
...
         for (ConstraintViolation<Entity> violation : violations) {
             System.out.println(violation.getMessageTemplate());
         }   
         //TODO: remove constraine violation: "The entity fields values are valid."
         Assertions.assertEquals(18, violations.size());
+        /*
+         * checking the minimal level of validation
+         */
+        violations.clear();
+        violations = emValidatorFactory.getValidator().validate(entityRecord.getEntity(), EntityFieldsMinimalValidatorGroup.class);
```
Comment (gsergiu) [inline entity-management-web/src/test/java/eu/europeana/entitymanagement/validation/EntityFieldsValidatorTest.java]: the mimimal validation must be implemented in own integration test

File: entity-management-web/src/test/java/eu/europeana/entitymanagement/validation/EntityFieldsValidatorTest.java [hunk 7]
```diff
@@ -43,33 +45,54 @@
...
+        Set<ConstraintViolation<Entity>> violations = emValidatorFactory.getValidator().validate(entityRecord.getEntity(), EntityFieldsCompleteValidatorGroup.class);
         for (ConstraintViolation<Entity> violation : violations) {
             System.out.println(violation.getMessageTemplate());
         }   
         //TODO: remove constraine violation: "The entity fields values are valid."
         Assertions.assertEquals(4, violations.size());
+        /*
+         * checking the minimal level of validation
+         */
+        violations.clear();
```
Comment (gsergiu) [inline entity-management-web/src/test/java/eu/europeana/entitymanagement/validation/EntityFieldsValidatorTest.java]: the nimimal validation must be implemented in own integration test

File: entity-management-web/src/main/java/eu/europeana/entitymanagement/batch/processor/EntityUpdateProcessor.java [hunk 8]
```diff
@@ -40,15 +41,15 @@ public EntityUpdateProcessor(EntityRecordService entityRecordService,
     public EntityRecord process(@NonNull EntityRecord entityRecord) throws EuropeanaApiException {
         //TODO: Validate entity metadata from Proxy Data Source
         emEntityFieldCleaner.cleanAndNormalize(entityRecord.getExternalProxy().getEntity());
-	entityRecordService.mergeEntity(entityRecord);
+        entityRecordService.mergeEntity(entityRecord);
```
Comment (gsergiu) [inline entity-management-web/src/main/java/eu/europeana/entitymanagement/batch/processor/EntityUpdateProcessor.java:46]: the minimal validation must not be integrated into the dereferencing step, but as own functionality in the EntityUpdateProcessor see step 4 in the specs https://docs.google.com/document/d/16k9PcCMFwl2LXjnnzotZRPc-QqM-Ar1D0VELHt4t_hA/edit#

**Review by gsergiu — CHANGES_REQUESTED** [review]
Please address the review comments

File: entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsCleaner.java (hunk 1 above)
Comment (SrdjanStevanetic) [reply entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsCleaner.java]: Ok.

File: entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsMinimalValidator.java (hunk 3 above)
Comment (SrdjanStevanetic) [reply entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsMinimalValidator.java:38]: Improved.

File: entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsCompleteValidator.java (hunk 2 above)
Comment (SrdjanStevanetic) [reply entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsCompleteValidator.java:207]: Merged.

File: entity-management-web/src/main/java/eu/europeana/entitymanagement/batch/processor/EntityDereferenceProcessor.java (hunk 4 above)
Comment (SrdjanStevanetic) [reply entity-management-web/src/main/java/eu/europeana/entitymanagement/batch/processor/EntityDereferenceProcessor.java:39]: Improved.

File: entity-management-web/src/main/java/eu/europeana/entitymanagement/batch/processor/EntityDereferenceProcessor.java (hunk 5 above)
Comment (SrdjanStevanetic) [reply entity-management-web/src/main/java/eu/europeana/entitymanagement/batch/processor/EntityDereferenceProcessor.java]: Improved.

File: entity-management-web/src/test/java/eu/europeana/entitymanagement/validation/EntityFieldsValidatorTest.java (hunk 6 above)
Comment (SrdjanStevanetic) [reply entity-management-web/src/test/java/eu/europeana/entitymanagement/validation/EntityFieldsValidatorTest.java]: Done.

File: entity-management-web/src/test/java/eu/europeana/entitymanagement/validation/EntityFieldsValidatorTest.java (hunk 7 above)
Comment (SrdjanStevanetic) [reply entity-management-web/src/test/java/eu/europeana/entitymanagement/validation/EntityFieldsValidatorTest.java]: Done.

File: entity-management-web/src/main/java/eu/europeana/entitymanagement/batch/processor/EntityUpdateProcessor.java (hunk 8 above)
Comment (SrdjanStevanetic) [reply entity-management-web/src/main/java/eu/europeana/entitymanagement/batch/processor/EntityUpdateProcessor.java:46]: Improved.

**Pull request merged**